mix.lock

```diff
@@ -1,92 +1,89 @@
   "phoenix_ecto": {:hex, :phoenix_ecto, "4.1.0", "a044d0756d0464c5a541b4a0bf4bcaf89bffcaf92468862408290682c73ae50d", [:mix], [{:ecto, "~> 3.0", [hex: :ecto, repo: "hexpm", optional: false]}, {:phoenix_html, "~> 2.9", [hex: :phoenix_html, repo: "hexpm", optional: true]}, {:plug, "~> 1.0", [hex: :plug, repo: "hexpm", optional: false]}], "hexpm", "c5e666a341ff104d0399d8f0e4ff094559b2fde13a5985d4cb5023b2c2ac558b"},
-  "phoenix_html": {:hex, :phoenix_html, "2.14.1", "7dabafadedb552db142aacbd1f11de1c0bbaa247f90c449ca549d5e30bbc66b4", [:mix], [{:plug, "~> 1.5", [hex: :plug, repo: "hexpm", optional: false]}], "hexpm", "536d5200ad37fecfe55b3241d90b7a8c3a2ca60cd012fc065f776324fa9ab0a9"},
+  "phoenix_html": {:hex, :phoenix_html, "2.14.3", "51f720d0d543e4e157ff06b65de38e13303d5778a7919bcc696599e5934271b8", [:mix], [{:plug, "~> 1.5", [hex: :plug, repo: "hexpm", optional: false]}], "hexpm", "efd697a7fff35a13eeeb6b43db884705cba353a1a41d127d118fda5f90c8e80f"},
   "phoenix_live_reload": {:hex, :phoenix_live_reload, "1.2.1", "274a4b07c4adbdd7785d45a8b0bb57634d0b4f45b18d2c508b26c0344bd59b8f", [:mix], [{:file_system, "~> 0.2.1 or ~> 0.3", [hex: :file_system, repo: "hexpm", optional: false]}, {:phoenix, "~> 1.4", [hex: :phoenix, repo: "hexpm", optional: false]}], "hexpm", "41b4103a2fa282cfd747d377233baf213c648fdcc7928f432937676532490eee"},
-  "phoenix_pubsub": {:hex, :phoenix_pubsub, "1.1.2", "496c303bdf1b2e98a9d26e89af5bba3ab487ba3a3735f74bf1f4064d2a845a3e", [:mix], [], "hexpm", "1f13f9f0f3e769a667a6b6828d29dec37497a082d195cc52dbef401a9b69bf38"},
-  "plug": {:hex, :plug, "1.10.4", "41eba7d1a2d671faaf531fa867645bd5a3dce0957d8e2a3f398ccff7d2ef017f", [:mix], [{:mime, "~> 1.0", [hex: :mime, repo: "hexpm", optional: false]}, {:plug_crypto, "~> 1.1.1 or ~> 1.2", [hex: :plug_crypto, repo: "hexpm", optional: false]}, {:telemetry, "~> 0.4", [hex: :telemetry, repo: "hexpm", optional: false]}], "hexpm", "ad1e233fe73d2eec56616568d260777b67f53148a999dc2d048f4eb9778fe4a0"},
+  "phoenix_pubsub": {:hex, :phoenix_pubsub, "2.1.1", "ba04e489ef03763bf28a17eb2eaddc2c20c6d217e2150a61e3298b0f4c2012b5", [:mix], [], "hexpm", "81367c6d1eea5878ad726be80808eb5a787a23dee699f96e72b1109c57cdd8d9"},
+  "plug": {:hex, :plug, "1.14.0", "ba4f558468f69cbd9f6b356d25443d0b796fbdc887e03fa89001384a9cac638f", [:mix], [{:mime, "~> 1.0 or ~> 2.0", [hex: :mime, repo: "hexpm", optional: false]}, {:plug_crypto, "~> 1.1.1 or ~> 1.2", [hex: :plug_crypto, repo: "hexpm", optional: false]}, {:telemetry, "~> 0.4.3 or ~> 1.0", [hex: :telemetry, repo: "hexpm", optional: false]}], "hexpm", "bf020432c7d4feb7b3af16a0c2701455cbbbb95e5b6866132cb09eb0c29adc14"},
```

Comment:
It seems that as of plug `1.14.0`, we cannot assign a `"host"` header with
```elixir
Conn.put_req_header(conn, "host", authority)
```
which we do in `PageController`: https://github.com/mozilla/reticulum/blob/dd14e634699d72c2640b459b31778551feed6e75/lib/ret_web/controllers/page_controller.ex#L720-L722

https://github.com/elixir-plug/plug/pull/1104

> The "host" header will be overridden by conn.host and should not be set with this method. Instead, do `%Plug.Conn{conn | host: value}`

https://hexdocs.pm/plug/Plug.Conn.html#put_req_header/3

I tried following the instructions above but have not been able to get `PageController.cors_proxy/2` working properly. I might not have my local environment configured correctly.

Reply:
It seems as though requesting a proxied resource results in rapid redirects that fail to resolve. (?)
![snip](https://user-images.githubusercontent.com/4072106/215299680-fd03b9bc-42be-4c50-b16e-f33f36c87258.png)

Reply:
OK. Despite the warnings in `Plug`, it seems like in this case we actually _want_ `conn.host` to be different from the `{"host", value}` in `conn.req_headers`:
```elixir
%Conn{ host: "hubs-proxy.local", req_headers: [{ "host", "uploads-prod.reticulum.io" }] }
```

Here is a fix:

```diff
modified   lib/ret_web/controllers/page_controller.ex
@@ -719,7 +719,7 @@ defmodule RetWeb.PageController do
         |> Map.put(:path_info, [])
         # Since we replaced the host with the IP address in ip_url above, we need to force the host
         # header back to the original authority so that the proxy destination does not reject our request
-        |> Conn.put_req_header("host", authority)
+        |> unsafe_put_req_header("host", authority)
         # Some domains disallow access from improper Origins
         |> Conn.delete_req_header("origin")
         |> ReverseProxyPlug.request(body, opts)
@@ -730,6 +730,14 @@ defmodule RetWeb.PageController do
     end
   end
 
+  # Bypass Plug.Conn.validate_req_header!/2 in order to set the "host" header
+  # https://github.com/elixir-plug/plug/blob/d7940552e365e0d3245bb78ca99429ea5c324f7d/lib/plug/conn.ex#L1887-L1892
+  # https://github.com/elixir-plug/plug/blob/d7940552e365e0d3245bb78ca99429ea5c324f7d/lib/plug/conn.ex#L785-L789
+  defp unsafe_put_req_header(%Conn{req_headers: headers} = conn, "host" = key, value)
+       when is_binary(value) do
+    %{conn | req_headers: headers |> List.delete(key) |> List.keystore(key, 0, {key, value})}
+  end
```

Reply:
It’s unfortunate that this deprecated behavior didn’t show up in the [Plug changelog](https://github.com/elixir-plug/plug/blob/main/CHANGELOG.md). I reviewed all the changelogs while upgrading.